From 3d0a31ca4423aa7a0c984ec83528daf177d90f6b Mon Sep 17 00:00:00 2001
From: kukutyin <attila.sagik@gmail.com>
Date: Wed, 27 May 2026 11:10:51 +0200
Subject: [PATCH] fix(ts_51_011): correct access conditions length and
 lifecycle decoding

- fix slicing: access_conditions now 3 bytes (not 2)
- implement proper LCS decoding per TS 102 221 / TS 31.101
- previous implementation misclassified multiple states
---
 pySim/ts_51_011.py | 20 ++++++++++++++++----
 1 file changed, 16 insertions(+), 4 deletions(-)

diff --git a/pySim/ts_51_011.py b/pySim/ts_51_011.py
index 89ff824..323696b 100644
--- a/pySim/ts_51_011.py
+++ b/pySim/ts_51_011.py
@@ -1303,13 +1303,25 @@ class CardProfileSIM(CardProfile):
                 record_len = resp_bin[14]
                 ret['file_descriptor']['record_len'] = record_len
                 ret['file_descriptor']['num_of_rec'] = ret['file_size'] // record_len
-            ret['access_conditions'] = b2h(resp_bin[8:10])
-            if resp_bin[11] & 0x01 == 0:
+            ret['access_conditions'] = b2h(resp_bin[8:11])
+            # Life cycle status
+            lcs = resp_bin[11]
+            if lcs & 0xF0:
+                ret['life_cycle_status_int'] = 'proprietary'
+            elif lcs == 0x00:
+                ret['life_cycle_status_int'] = 'no_information_given'
+            elif lcs == 0x01:
+                ret['life_cycle_status_int'] = 'creation_state'
+            elif lcs == 0x03:
+                ret['life_cycle_status_int'] = 'initialization_state'
+            elif lcs in (0x05, 0x07):
                 ret['life_cycle_status_int'] = 'operational_activated'
-            elif resp_bin[11] & 0x04:
+            elif lcs in (0x04, 0x06):
                 ret['life_cycle_status_int'] = 'operational_deactivated'
+            elif 0x0C <= lcs <= 0x0F:
+                ret['life_cycle_status_int'] = 'termination_state'
             else:
-                ret['life_cycle_status_int'] = 'terminated'
+                ret['life_cycle_status_int'] = 'rfu'
         return ret
 
     @classmethod
-- 
2.39.5

